extensions/amp-link-rewriter/0.1/link-shifter.js

```diff
@@ -0,0 +1,227 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  CTX_ATTR_NAME,
+  CTX_ATTR_VALUE,
+  NS_DATA_PH,
+  PREFIX_DATA_ATTR,
+  WL_ANCHOR_ATTR,
+} from './constants';
+import {getConfigOpts} from './config-options';
+
+export class LinkShifter {
+  /**
+   * @param {!AmpElement} ampElement
+   * @param {?../../../src/service/viewer-impl.Viewer} viewer
+   */
+  constructor(ampElement, viewer) {
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = viewer;
+
+    /** @private {?Event} */
+    this.event_ = null;
+
+    /** @private {?Object} */
+    this.configOpts_ = getConfigOpts(ampElement);
+
+    /** @private {boolean|number|string} */
+    this.ctxAttrValue_ = CTX_ATTR_VALUE().toString();
+
+    /** @private {?RegExp} */
+    this.regexDomainUrl_ = /^https?:\/\/(www\.)?([^\/:]*)(:\d+)?(\/.*)?$/;
+  }
+
+  /**
+   * @param {!Event} event
+   */
+  clickHandler(event) {
+    this.event_ = event;
+    const htmlElement = event.srcElement;
+    const trimmedDomain = this.viewer_.win.document.domain
+        .replace(/(www\.)?(.*)/, '$2');
+
+    // avoid firefox to trigger the event twice (just caution)
+    if ((this.event_.type !== 'contextmenu') && (this.event_.button === 2)) {
+      return;
+    }
+
+    if (this.wasShifted_(htmlElement)) {
+      return;
+    }
+
+    if (this.isInternalLink(htmlElement, trimmedDomain)) {
+      return;
+    }
+
+    this.setTrackingUrl_(htmlElement);
+  }
+
+  /**
+   * Check if the anchor element was already shifted
+   * @param {?Element} htmlElement
+   * @return {boolean}
+   * @private
+   */
+  wasShifted_(htmlElement) {
+    const ctxAttrValue = this.ctxAttrValue_;
+    return (htmlElement.hasAttribute(CTX_ATTR_NAME)) &&
+        (htmlElement.getAttribute(CTX_ATTR_NAME) === ctxAttrValue);
+  }
+
+  /**
+   * Check if the anchor element leads to an internal link
+   * @param {?Element} htmlElement
+   * @param {?string} trimmedDomain
+   * @return {boolean}
+   */
+  isInternalLink(htmlElement, trimmedDomain) {
+    const href = htmlElement.getAttribute('href');
+
+    if (!(href && this.regexDomainUrl_.test(href) &&
+            RegExp.$2 !== trimmedDomain)
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   *
+   * @param {?Element} htmlElement
+   */
+  setTrackingUrl_(htmlElement) {
+
+    const oldValHref = htmlElement.getAttribute('href');
+
+    this.viewer_.getReferrerUrl().then(referrerUrl => {
+      const pageAttributes = {
```

Comment:
we can read and set those global vars in constructor, and get it merged with user provided var.

```
constructor() {
  this.vars_ = ... // assign referrer and location, merge with user provided vars
}
```

and here,  you do:

```
this.vars_.then(vars => {
  ...
  this.replacePlaceHolders(htmlElement, vars);
  ...
});
```